{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false
   },
   "source": [
    "### MEDC0106: Bioinformatics in Applied Biomedical Science\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"../../resources/static/Banner.png\" alt=\"MEDC0106 Banner\" width=\"90%\"/>\n",
    "  <br>\n",
    "</p>\n",
    "\n",
    "---------------------------------------------------------------\n",
    "\n",
    "# 09 - Introduction to Biopython - Sequences\n",
    "\n",
    "*Written by:* Mateusz Kaczy≈Ñski\n",
    "\n",
    "**This notebook provides a general introduction to Biopython library and provides in depth look at dealing with the sequences - the cornerstone of bioinformatics.**\n",
    "\n",
    "\n",
    "### What is Biopython?\n",
    "\n",
    "**Biopython** is a popular open-source toolbox for computational biology and bioinformatics. It contains tools and connectors for various resources and provides functions for running common bioinformatics tasks and parsing common bioinformatics data formats.\n",
    "\n",
    "\n",
    "### Why Biopython?\n",
    "**Biopython** is a de-facto standard for accessing a lot of the databases and tools making it easier to share your work and results. It simplifies the use the libraries written in other languages and technologies as wekk as those hosted online. While it is possible to use those directly directly - Biopython simplifies the process and provides the components you. \n",
    " \n",
    "\n",
    "\n",
    "## Contents\n",
    "\n",
    "\n",
    "1. [Sequence basics](#Sequence-basics)\n",
    "2. [Transcription and translation](#Transcription-and-translation)\n",
    "3. [Alignment](#Alignment)\n",
    "4. [Downloading and reading FASTA files](#Downloading-and-reading-fasta-files)\n",
    "5. [Discussion](#Discussion)\n",
    "-----\n",
    "\n",
    "#### Extra Resources:\n",
    "\n",
    "- [Official Biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html) - A comprehensive guide to what the library capabilities.\n",
    "- [Rosalind](http://rosalind.info) - A bioinformatics learning platform that includes exercises.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence basics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Module Bio version 1.79\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "print(\"Module\", Bio.__name__, \"version\", Bio.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Definiting a sentence\n",
    "from Bio.Seq import Seq\n",
    "dna_sequence = Seq(\"ACTG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence compared with the original string True\n",
      "The type of dna_sequence <class 'Bio.Seq.Seq'>\n"
     ]
    }
   ],
   "source": [
    "print(\"The sequence compared with the original string:\", dna_sequence == \"ACTG\")\n",
    "print(\"The type of dna_sequence:\", type(dna_sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Seq in module Bio.Seq object:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data, length=None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data, length=None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_seq = Seq(None, 20)\n",
      " |      >>> my_undefined_seq\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_seq)\n",
      " |      20\n",
      " |      >>> print(my_undefined_seq)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |  \n",
      " |  complement(self)\n",
      " |      Return the complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      This method is intended for use with DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAG\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAG')\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGCTATC')\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-GD\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-GD')\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGctaT-CH')\n",
      " |      \n",
      " |      Note in the above example, ambiguous character D denotes\n",
      " |      G, A or T so its complement is H (for C, T or A).\n",
      " |      \n",
      " |      Note that if the sequence contains neither T nor U, we\n",
      " |      assume it is DNA and map any A character to T:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      >>> Seq(\"CGAT\").complement()\n",
      " |      Seq('GCTA')\n",
      " |      \n",
      " |      If you actually have RNA, this currently works but we\n",
      " |      may deprecate this later. We recommend using the new\n",
      " |      complement_rna method instead:\n",
      " |      \n",
      " |      >>> Seq(\"CGAU\").complement()\n",
      " |      Seq('GCUA')\n",
      " |      >>> Seq(\"CGAU\").complement_rna()\n",
      " |      Seq('GCUA')\n",
      " |      \n",
      " |      If the sequence contains both T and U, an exception is\n",
      " |      raised:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Mixed RNA/DNA found\n",
      " |      \n",
      " |      Trying to complement a protein sequence gives a meaningless\n",
      " |      sequence:\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\")\n",
      " |      >>> my_protein.complement()\n",
      " |      Seq('KTIBKCY')\n",
      " |      \n",
      " |      Here \"M\" was interpreted as the IUPAC ambiguity code for\n",
      " |      \"A\" or \"C\", with complement \"K\" for \"T\" or \"G\". Likewise\n",
      " |      \"A\" has complement \"T\". The letter \"I\" has no defined\n",
      " |      meaning under the IUPAC convention, and is unchanged.\n",
      " |  \n",
      " |  encode(self, encoding='utf-8', errors='strict')\n",
      " |      Return an encoded version of the sequence as a bytes object.\n",
      " |      \n",
      " |      The Seq object aims to match the interface of a Python string.\n",
      " |      \n",
      " |      This is essentially to save you doing str(my_seq).encode() when\n",
      " |      you need a bytes string, for example for computing a hash:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"ACGT\").encode(\"ascii\")\n",
      " |      b'ACGT'\n",
      " |  \n",
      " |  reverse_complement(self)\n",
      " |      Return the reverse complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      This method is intended for use with DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAGNR\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAGNR')\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('YNCTATCGGGGG')\n",
      " |      \n",
      " |      Note in the above example, since R = G or A, its complement\n",
      " |      is Y (which denotes C or T).\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-G\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-G')\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('C-TatcGGGGG')\n",
      " |      \n",
      " |      As discussed for the complement method, if the sequence\n",
      " |      contains neither T nor U, is is assumed to be DNA and\n",
      " |      will map any letter A to T.\n",
      " |      \n",
      " |      If you are dealing with RNA you should use the new\n",
      " |      reverse_complement_rna method instead\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()  # defaults to DNA\n",
      " |      Seq('TCG')\n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      If the sequence contains both T and U, an exception is raised:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Mixed RNA/DNA found\n",
      " |      \n",
      " |      Trying to reverse complement a protein sequence will give\n",
      " |      a meaningless sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGR\")\n",
      " |      >>> my_protein.reverse_complement()\n",
      " |      Seq('YCKBITK')\n",
      " |      \n",
      " |      Here \"M\" was interpretted as the IUPAC ambiguity code for\n",
      " |      \"A\" or \"C\", with complement \"K\" for \"T\" or \"G\" - and so on.\n",
      " |  \n",
      " |  tomutable(self)\n",
      " |      Return the full sequence as a MutableSeq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAAL')\n",
      " |      >>> my_seq.tomutable()\n",
      " |      MutableSeq('MKQHKAMIVALIVICITAVVAAL')\n",
      " |  \n",
      " |  ungap(self, gap='-')\n",
      " |      Return a copy of the sequence without the gap character(s) (OBSOLETE).\n",
      " |      \n",
      " |      The gap character now defaults to the minus sign, and can only\n",
      " |      be specified via the method argument. This is no longer possible\n",
      " |      via the sequence's alphabet (as was possible up to Biopython 1.77):\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\")\n",
      " |      >>> my_dna\n",
      " |      Seq('-ATA--TGAAAT-TTGAAAA')\n",
      " |      >>> my_dna.ungap(\"-\")\n",
      " |      Seq('ATATGAAATTTGAAAA')\n",
      " |      \n",
      " |      This method is OBSOLETE; please use my_dna.replace(gap, \"\") instead.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 which ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip(inplace=False)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at positon 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at positon 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip(inplace=False)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip(inplace=False)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(dna_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of the sentence: 4\n",
      "AC found in the sequence: True\n",
      "Sequence ends with 'TG': True\n",
      "The first base: A\n",
      "The last base: G\n",
      "Slicing - first two bases: AC\n",
      "How many Adanines are in the sequence: 1\n"
     ]
    }
   ],
   "source": [
    "#String-like behaviour\n",
    "print(\"The length of the sentence:\", len(dna_sequence))\n",
    "print(\"AC found in the sequence:\", \"AC\" in dna_sequence)\n",
    "print(\"Sequence ends with 'TG':\", dna_sequence.endswith(\"TG\"))\n",
    "print(\"The first base:\", dna_sequence[0])\n",
    "print(\"The last base:\", dna_sequence[-1])\n",
    "print(\"Slicing - first two bases:\", dna_sequence[0:2])\n",
    "print(\"How many Adanines are in the sequence:\", dna_sequence.count(\"A\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upper case sequence equal to mixed case sentence:        False\n",
      "How about if we change the original to lower case?:      False\n",
      "How about moving both to the same case? Are they equal?: True\n"
     ]
    }
   ],
   "source": [
    "#Case insensitivity\n",
    "dna_sequence_mixedcase = Seq(\"acTG\")\n",
    "print(\"Upper case sequence equal to mixed case sentence:       \", \n",
    "      dna_sequence == dna_sequence_mixedcase)\n",
    "print(\"How about if we change the original to lower case?:     \", \n",
    "      dna_sequence.lower() == dna_sequence_mixedcase)\n",
    "print(\"How about moving both to the same case? Are they equal?:\", \n",
    "      dna_sequence.lower() == dna_sequence_mixedcase.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ACTGTGG\n"
     ]
    }
   ],
   "source": [
    "# Extending the sequence\n",
    "print(\"\", dna_sequence + \"TGG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcription and translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sentence with gaps   : ACTG-ACTG-ACTG\n",
      "Sentence without gaps: ACTGACTGACTG\n"
     ]
    }
   ],
   "source": [
    "#Ungap \n",
    "sequence_with_gaps = Seq(\"-\".join(3*[str(dna_sequence)]))\n",
    "print(\"Sentence with gaps   :\", sequence_with_gaps)\n",
    "print(\"Sentence without gaps:\", sequence_with_gaps.ungap())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The guanine-cytosine content of the sequence (%): 50.0\n"
     ]
    }
   ],
   "source": [
    "# GC content count\n",
    "from Bio.SeqUtils import GC\n",
    "print(\"The guanine-cytosine content of the sequence (%):\", GC(dna_sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence:           ACTG\n",
      "Complement:         TGAC\n",
      "Reverse:            GTCA\n",
      "Reverse complement: CAGT\n"
     ]
    }
   ],
   "source": [
    "# Complement\n",
    "print(\"Sequence:          \", dna_sequence)\n",
    "print(\"Complement:        \", dna_sequence.complement())\n",
    "print(\"Reverse:           \", Seq(\"\".join(reversed(dna_sequence))))\n",
    "print(\"Reverse complement:\", dna_sequence.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence:               ACTG\n",
      "Transcribed:            ACUG\n",
      "Reverse complement RNA: CAGU\n"
     ]
    }
   ],
   "source": [
    "# DNA to RNA\n",
    "print(\"Sequence:              \", dna_sequence)\n",
    "print(\"Transcribed:           \", dna_sequence.transcribe())\n",
    "print(\"Reverse complement RNA:\", dna_sequence.reverse_complement_rna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcibed:       ACUG\n",
      "Back transcribed: ACTG\n"
     ]
    }
   ],
   "source": [
    "# From RNA to DNA \n",
    "transcribed = dna_sequence.transcribe()\n",
    "print(\"Transcibed:      \", transcribed)\n",
    "print(\"Back transcribed:\", transcribed.back_transcribe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TRR')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To Protein\n",
    "sequence = Seq(\"ACGCGACGA\")\n",
    "sequence.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcription directly from 'ACGT' string, without using Seq object: ACGU\n"
     ]
    }
   ],
   "source": [
    "# Operations on strings directly.\n",
    "from Bio.Seq import transcribe\n",
    "print(\"Transcription directly from 'ACGT' string, without using Seq object:\", transcribe(\"ACGT\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment(seqA='ACGT-', seqB='ACG-C', score=3.0, start=0, end=5)\n",
      "Alignment(seqA='ACGT', seqB='ACGC', score=3.0, start=0, end=4)\n"
     ]
    }
   ],
   "source": [
    "# Two-sequence alignment\n",
    "from Bio import pairwise2\n",
    "\n",
    "# Notice we can supply either Seq's or strings directly.\n",
    "global_alignments = pairwise2.align.globalxx(Seq(\"ACGT\"), \"ACGC\")\n",
    "for ga in global_alignments:\n",
    "    print(ga)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACGT-\n",
      "|||  \n",
      "ACG-C\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print the first available alignment. The * converts the Alignment objects into a list of parameters required by the format_alignment function.\n",
    "print(pairwise2.format_alignment(*global_alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 ACG\n",
      "  |||\n",
      "1 ACG\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Similarly we can use local alignments.\n",
    "local_alignments = pairwise2.align.localxx(Seq(\"ACGT\"), \"ACGC\")\n",
    "print(pairwise2.format_alignment(*local_alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The simple consensus: ACGTACGX\n"
     ]
    }
   ],
   "source": [
    "# Multiple sequence alignment\n",
    "from Bio.Align import MultipleSeqAlignment, AlignInfo\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "seqs = [\n",
    "    \"ACGTACGT\",\n",
    "    \"ACGTGCGC\",\n",
    "    \"ACGTA--T\",\n",
    "    \"CCGTACGG\",\n",
    "    \"A-GTACCC\",\n",
    "    \"ACGTA--T\",\n",
    "    \"CTG-ACG-\",\n",
    "    \"AGGTACG-\"\n",
    "]\n",
    "aligned = MultipleSeqAlignment([SeqRecord(Seq(s)) for s in seqs])\n",
    "align_info = AlignInfo.SummaryInfo(aligned)\n",
    "print(\"The simple consensus:\", align_info.dumb_consensus())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment score for particular bases at a given position.\n",
      "Dumb consensus on the y axis up-to-down, bases on x-axis.\n",
      "    A   C   G   T\n",
      "A  6.0 2.0 0.0 0.0\n",
      "C  0.0 5.0 1.0 1.0\n",
      "G  0.0 0.0 8.0 0.0\n",
      "T  0.0 0.0 0.0 7.0\n",
      "A  7.0 0.0 1.0 0.0\n",
      "C  0.0 6.0 0.0 0.0\n",
      "G  0.0 1.0 5.0 0.0\n",
      "X  0.0 2.0 1.0 3.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Alignment score for particular bases at a given position.\")\n",
    "print(\"Dumb consensus on the y axis up-to-down, bases on x-axis.\")\n",
    "print(align_info.pos_specific_score_matrix())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading and reading fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded file to:  data/ENSG00000001626.fasta\n",
      "\n",
      "Download metadata and statistics:\n",
      "Vary: Content-Type\n",
      "Vary: Origin\n",
      "Content-Type: text/x-fasta; charset=UTF-8\n",
      "Date: Tue, 23 Nov 2021 13:40:33 GMT\n",
      "X-RateLimit-Limit: 55000\n",
      "X-RateLimit-Reset: 1167\n",
      "X-Runtime: 0.280467\n",
      "Connection: close\n",
      "X-RateLimit-Period: 3600\n",
      "X-RateLimit-Remaining: 54999\n",
      "Content-Length: 436062\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# File downloads\n",
    "from urllib.request import urlretrieve \n",
    "# CFTR gene - homo sapiens - fasta URL from Ensembl / P13569.\n",
    "CFTR_FASTA_path = \"data/ENSG00000001626.fasta\"\n",
    "result_location, http_response = urlretrieve(\"https://rest.ensembl.org/sequence/id/ENSG00000001626.fasta\", CFTR_FASTA_path)\n",
    "print(\"Downloaded file to: \", result_location)\n",
    "\n",
    "print(\"\\nDownload metadata and statistics:\")\n",
    "print(http_response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ENSG00000001626.16\n",
      "Name: ENSG00000001626.16\n",
      "Description: ENSG00000001626.16 chromosome:GRCh38:7:117287120:117715971:1\n",
      "Number of features: 0\n",
      "Seq('AGGCGGATCACAAGTTCATGAGATCGAGACCATCTTGGCCAACATGGTGAGACC...ACA')\n",
      "Sequence length: 428852\n",
      "GC content: 36.895245912342716\n"
     ]
    }
   ],
   "source": [
    "# Reading FASTA files. Note that each file can contain multiple records.\n",
    "from Bio import SeqIO\n",
    "for record in SeqIO.parse(CFTR_FASTA_path, \"fasta\"):\n",
    "    print(record)\n",
    "    print(\"Sequence length:\", len(record.seq))\n",
    "    print(\"GC content:\", GC(record.seq))"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
