{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "### MEDC0106: Bioinformatics in Applied Biomedical Science\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"../../resources/static/Banner.png\" alt=\"MEDC0106 Banner\" width=\"90%\"/>\n",
    "  <br>\n",
    "</p>\n",
    "\n",
    "---------------------------------------------------------------\n",
    "\n",
    "# 09 - Introduction to Biopython - Sequences\n",
    "\n",
    "*Written by:* Mateusz Kaczy≈Ñski\n",
    "\n",
    "**This notebook provides a general introduction to Biopython library and provides an in-depth look at dealing with the sequences - the cornerstone of bioinformatics.**\n",
    "\n",
    "\n",
    "### What is Biopython?\n",
    "\n",
    "**Biopython** is a popular open-source toolbox for computational biology and bioinformatics. It contains tools and connectors for various resources and provides functions for running common bioinformatics tasks and parsing common bioinformatics data formats.\n",
    "\n",
    "\n",
    "### Why Biopython?\n",
    "**Biopython** is a de-facto standard for accessing a lot of the databases and tools making it easier to share your work and results. It simplifies the use of the libraries written in other languages and technologies as well as those hosted online. While it is possible to use those directly - Biopython simplifies the process and provides the components you. \n",
    " \n",
    "\n",
    "\n",
    "## Contents\n",
    "\n",
    "\n",
    "1. [Sequence basics](#Sequence-basics)\n",
    "2. [Transcription and translation](#Transcription-and-translation)\n",
    "3. [Alignment](#Alignment)\n",
    "4. [Downloading and reading FASTA files](#Downloading-and-reading-fasta-files)\n",
    "5. [Discussion](#Discussion)\n",
    "-----\n",
    "\n",
    "#### Extra Resources:\n",
    "\n",
    "- [Official Biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html) - A comprehensive guide to the library capabilities.\n",
    "- [Biopython API documentation](https://biopython.org/docs/latest/api/index.html) - a long, detailed list of all methods and connectors provided by Biopython.\n",
    "- [Rosalind](http://rosalind.info) - A bioinformatics learning platform that includes exercises.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "#### Importing Biopython\n",
    "While installed, **Biopython** can be included in the project by importing it as `Bio` module. It is not that uncommon for the library to be called differently than the imported name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Module Bio version 1.79\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "print(\"Module\", Bio.__name__, \"version\", Bio.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "## Sequence basics\n",
    "Managing sequences of DNA, RNA or peptides are a core part of any bioinformatics workflow. Being able to understand, transform and use them is one of the most beneficial uses of computers in biology.\n",
    "\n",
    "Here is how we define a simple sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "dna_sequence = Seq(\"ACTG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "You can see that the sequence can be created from a simple `string` structure representing the ordered items (in this case - DNA). However, it is no longer a simple data type.\n",
    "\n",
    "Let's take a closer look at the Python representation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence compared with the original string: True\n",
      "The type of dna_sequence: <class 'Bio.Seq.Seq'>\n"
     ]
    }
   ],
   "source": [
    "print(\"The sequence compared with the original string:\", dna_sequence == \"ACTG\")\n",
    "print(\"The type of dna_sequence:\", type(dna_sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "### Objects and classes\n",
    "We can see that, while we can compare it with the original `string` object without an issue, the type of the variable is `Seq`.\n",
    "\n",
    "`dna_sequence` is an `object` of `class` `Seq`. \n",
    "    \n",
    "Let's unfold those concepts:\n",
    "  - `object` - a fundamental building block of a program - it can contain properties and functions. In Python almost anything is an `object`.\n",
    "  - `class` - a blueprint for creating new objects, provides a way to mix data and functions\n",
    "  - `Seq` - a particular `class` defined in `Bio.Seq` module.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "When in doubt, we can use `help` function to help us understand how to use `Seq`s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Seq in module Bio.Seq object:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data, length=None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data, length=None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_seq = Seq(None, 20)\n",
      " |      >>> my_undefined_seq\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_seq)\n",
      " |      20\n",
      " |      >>> print(my_undefined_seq)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |  \n",
      " |  complement(self)\n",
      " |      Return the complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      This method is intended for use with DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAG\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAG')\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGCTATC')\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-GD\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-GD')\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGctaT-CH')\n",
      " |      \n",
      " |      Note in the above example, ambiguous character D denotes\n",
      " |      G, A or T so its complement is H (for C, T or A).\n",
      " |      \n",
      " |      Note that if the sequence contains neither T nor U, we\n",
      " |      assume it is DNA and map any A character to T:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      >>> Seq(\"CGAT\").complement()\n",
      " |      Seq('GCTA')\n",
      " |      \n",
      " |      If you actually have RNA, this currently works but we\n",
      " |      may deprecate this later. We recommend using the new\n",
      " |      complement_rna method instead:\n",
      " |      \n",
      " |      >>> Seq(\"CGAU\").complement()\n",
      " |      Seq('GCUA')\n",
      " |      >>> Seq(\"CGAU\").complement_rna()\n",
      " |      Seq('GCUA')\n",
      " |      \n",
      " |      If the sequence contains both T and U, an exception is\n",
      " |      raised:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Mixed RNA/DNA found\n",
      " |      \n",
      " |      Trying to complement a protein sequence gives a meaningless\n",
      " |      sequence:\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\")\n",
      " |      >>> my_protein.complement()\n",
      " |      Seq('KTIBKCY')\n",
      " |      \n",
      " |      Here \"M\" was interpreted as the IUPAC ambiguity code for\n",
      " |      \"A\" or \"C\", with complement \"K\" for \"T\" or \"G\". Likewise\n",
      " |      \"A\" has complement \"T\". The letter \"I\" has no defined\n",
      " |      meaning under the IUPAC convention, and is unchanged.\n",
      " |  \n",
      " |  encode(self, encoding='utf-8', errors='strict')\n",
      " |      Return an encoded version of the sequence as a bytes object.\n",
      " |      \n",
      " |      The Seq object aims to match the interface of a Python string.\n",
      " |      \n",
      " |      This is essentially to save you doing str(my_seq).encode() when\n",
      " |      you need a bytes string, for example for computing a hash:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"ACGT\").encode(\"ascii\")\n",
      " |      b'ACGT'\n",
      " |  \n",
      " |  reverse_complement(self)\n",
      " |      Return the reverse complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      This method is intended for use with DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAGNR\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAGNR')\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('YNCTATCGGGGG')\n",
      " |      \n",
      " |      Note in the above example, since R = G or A, its complement\n",
      " |      is Y (which denotes C or T).\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-G\")\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-G')\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('C-TatcGGGGG')\n",
      " |      \n",
      " |      As discussed for the complement method, if the sequence\n",
      " |      contains neither T nor U, is is assumed to be DNA and\n",
      " |      will map any letter A to T.\n",
      " |      \n",
      " |      If you are dealing with RNA you should use the new\n",
      " |      reverse_complement_rna method instead\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()  # defaults to DNA\n",
      " |      Seq('TCG')\n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      If the sequence contains both T and U, an exception is raised:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Mixed RNA/DNA found\n",
      " |      \n",
      " |      Trying to reverse complement a protein sequence will give\n",
      " |      a meaningless sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGR\")\n",
      " |      >>> my_protein.reverse_complement()\n",
      " |      Seq('YCKBITK')\n",
      " |      \n",
      " |      Here \"M\" was interpretted as the IUPAC ambiguity code for\n",
      " |      \"A\" or \"C\", with complement \"K\" for \"T\" or \"G\" - and so on.\n",
      " |  \n",
      " |  tomutable(self)\n",
      " |      Return the full sequence as a MutableSeq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAAL')\n",
      " |      >>> my_seq.tomutable()\n",
      " |      MutableSeq('MKQHKAMIVALIVICITAVVAAL')\n",
      " |  \n",
      " |  ungap(self, gap='-')\n",
      " |      Return a copy of the sequence without the gap character(s) (OBSOLETE).\n",
      " |      \n",
      " |      The gap character now defaults to the minus sign, and can only\n",
      " |      be specified via the method argument. This is no longer possible\n",
      " |      via the sequence's alphabet (as was possible up to Biopython 1.77):\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\")\n",
      " |      >>> my_dna\n",
      " |      Seq('-ATA--TGAAAT-TTGAAAA')\n",
      " |      >>> my_dna.ungap(\"-\")\n",
      " |      Seq('ATATGAAATTTGAAAA')\n",
      " |      \n",
      " |      This method is OBSOLETE; please use my_dna.replace(gap, \"\") instead.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 which ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip(inplace=False)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement()\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at positon 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at positon 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip(inplace=False)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip(inplace=False)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(dna_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "### String-like behaviour\n",
    "In the first workshop, we became familiar with how Python `string` - simple text representation objects behave. \n",
    "\n",
    "`Seq` class is built to follow the same methods to help with interchanging the information between the two types of data and present a more familiar interface to the user.\n",
    "\n",
    "Let's verify this with some common methods, including indexing and slicing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of the sentence: 4\n",
      "AC found in the sequence:   True\n",
      "Sequence ends with 'TG':    True\n",
      "The first base:             A\n",
      "The last base:              G\n",
      "Slicing - first two bases:  AC\n",
      "How many Adenines are in the sequence: 1\n"
     ]
    }
   ],
   "source": [
    "print(\"The length of the sentence:\", len(dna_sequence))\n",
    "print(\"AC found in the sequence:  \", \"AC\" in dna_sequence)\n",
    "print(\"Sequence ends with 'TG':   \", dna_sequence.endswith(\"TG\"))\n",
    "print(\"The first base:            \", dna_sequence[0])\n",
    "print(\"The last base:             \", dna_sequence[-1])\n",
    "print(\"Slicing - first two bases: \", dna_sequence[0:2])\n",
    "print(\"How many Adenines are in the sequence:\", dna_sequence.count(\"A\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "What about case sensitivity? Does it matter if we represent Adenine with `a` or `A`? \n",
    "\n",
    "Turns out that yes - it matters a lot. Due to the recent changes in **Biopython** (removal of `Alphabet` concept), there is no easy way to verify or enforce the content of `Seq` objects. Users are required to perform their own verification.\n",
    "\n",
    "Let's have a closer look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upper case sequence equal to mixed case sentence:        False\n",
      "How about if we change the original to lower case?:      False\n",
      "How about moving both to the same case? Are they equal?: True\n"
     ]
    }
   ],
   "source": [
    "dna_sequence_mixedcase = Seq(\"acTG\")\n",
    "print(\"Upper case sequence equal to mixed case sentence:       \", \n",
    "      dna_sequence == dna_sequence_mixedcase)\n",
    "print(\"How about if we change the original to lower case?:     \", \n",
    "      dna_sequence.lower() == dna_sequence_mixedcase)\n",
    "print(\"How about moving both to the same case? Are they equal?:\", \n",
    "      dna_sequence.lower() == dna_sequence_mixedcase.lower())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "To extend the sequence, we can use the familiar `string` syntax. \n",
    "\n",
    "Again, it does not matter if you use `string` or `Seq` class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extended with another sequence: ACTGTAA\n",
      "Extended with a string        : ACTGTGG\n"
     ]
    }
   ],
   "source": [
    "print(\"Extended with another sequence:\", dna_sequence + Seq(\"TAA\"))\n",
    "print(\"Extended with a string        :\", dna_sequence + \"TGG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "### Example sequence-specific functions \n",
    "In addition to `string`-like behaviour, `Seq` objects contain methods applicable in their specific domain. \n",
    "\n",
    "For the cases where there are gaps, `ungap` method helps to deal with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sentence with gaps   : ACTG-ACTG-ACTG\n",
      "Sentence without gaps: ACTGACTGACTG\n"
     ]
    }
   ],
   "source": [
    "sequence_with_gaps = Seq(\"-\".join(3*[str(dna_sequence)]))\n",
    "print(\"Sentence with gaps   :\", sequence_with_gaps)\n",
    "print(\"Sentence without gaps:\", sequence_with_gaps.ungap())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "**Biopython** contains various tools that can be used for sequence analysis, for example, GC content calculation can be imported like so. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The guanine-cytosine content of the sequence (%): 50.0\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqUtils import GC\n",
    "print(\"The guanine-cytosine content of the sequence (%):\", GC(dna_sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "## Transcription and translation\n",
    "One of the most powerful features of **Biopython** is switching from one representation to another, converting between seqeunce types.\n",
    "\n",
    "Let's take a look at how to generate complementary and reverse sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence:           ACTG\n",
      "Complement:         TGAC\n",
      "Reverse:            GTCA\n",
      "Reverse complement: CAGT\n"
     ]
    }
   ],
   "source": [
    "print(\"Sequence:          \", dna_sequence)\n",
    "print(\"Complement:        \", dna_sequence.complement())\n",
    "print(\"Reverse:           \", Seq(\"\".join(reversed(dna_sequence))))\n",
    "print(\"Reverse complement:\", dna_sequence.reverse_complement())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "Simple DNA-to-RNA transcription can be achieved with the following methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence:               ACTG\n",
      "Transcribed:            ACUG\n",
      "Reverse complement RNA: CAGU\n"
     ]
    }
   ],
   "source": [
    "print(\"Sequence:              \", dna_sequence)\n",
    "print(\"Transcribed:           \", dna_sequence.transcribe())\n",
    "print(\"Reverse complement RNA:\", dna_sequence.reverse_complement_rna())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "The reversal of the process - getting from RNA to DNA is also supported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcibed:       ACUG\n",
      "Back transcribed: ACTG\n"
     ]
    }
   ],
   "source": [
    "transcribed = dna_sequence.transcribe()\n",
    "print(\"Transcibed:      \", transcribed)\n",
    "print(\"Back transcribed:\", transcribed.back_transcribe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "**Biopython** makes translation from RNA (or directly DNA) to aminoacid sequence a very easy task. \n",
    "\n",
    "*Note that different codon tables are also supported.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TRR')"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence = Seq(\"ACGCGACGA\")\n",
    "sequence.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "### Using `Seq` methods directly on `string`s. \n",
    "It is possible to ignore `Seq` class abstraction and import and execute relevant methods directly on simple `string`s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcription directly from 'ACGT' string, without using Seq object: ACGU\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import transcribe\n",
    "print(\"Transcription directly from 'ACGT' string, without using Seq object:\", transcribe(\"ACGT\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "## Alignment\n",
    "\n",
    "One of the most common uses of bioinformatics tools is sequence alignment. \n",
    "**Biopython** provides support for various command-line and online alignment tools. The details can be found [here](https://biopython.org/docs/latest/api/Bio.Align.Applications.html). \n",
    "\n",
    "For this section of this workshop, we will use the built-in, out-of-the-box alignment method contained in `pairwise2` module. It allows aligning 2 sequences at a time.\n",
    "\n",
    "We will first import it, run a global (full sequence) alignment and present relevant results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment(seqA='ACGT-', seqB='ACG-C', score=3.0, start=0, end=5)\n",
      "Alignment(seqA='ACGT', seqB='ACGC', score=3.0, start=0, end=4)\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "# Notice we can supply either Seq's or strings directly.\n",
    "global_alignments = pairwise2.align.globalxx(Seq(\"ACGT\"), \"ACGC\")\n",
    "for ga in global_alignments:\n",
    "    print(ga)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "This representation is not really user friendly. \n",
    "\n",
    "Let's try to make it look better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACGT-\n",
      "|||  \n",
      "ACG-C\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print the first available alignment. \n",
    "# The * converts the Alignment objects into a list of parameters required by the format_alignment function.\n",
    "print(pairwise2.format_alignment(*global_alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "Similar to the global alignment, we can use `pairwise2` to perform local alignment of 2 sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 ACG\n",
      "  |||\n",
      "1 ACG\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "local_alignments = pairwise2.align.localxx(Seq(\"ACGT\"), \"ACGC\")\n",
    "print(pairwise2.format_alignment(*local_alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "Multiple sequence alignment is suitable for aligning 3 or more sequences at a time. \n",
    "\n",
    "*Note: for MSA, sequences are expected to be of the same length.* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import MultipleSeqAlignment, AlignInfo\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "seqs = [\n",
    "    \"ACGTACGT\",\n",
    "    \"ACGTGCGC\",\n",
    "    \"ACGTA--T\",\n",
    "    \"CCGTACGG\",\n",
    "    \"A-GTACCC\",\n",
    "    \"ACGTA--T\",\n",
    "    \"CTG-ACG-\",\n",
    "    \"AGGTACG-\"\n",
    "]\n",
    "aligned = MultipleSeqAlignment([SeqRecord(Seq(s)) for s in seqs])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "Aligning sequences does not represent too much value on its own. \n",
    "\n",
    "We may want to learn what statistics can be retrieved from it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on SummaryInfo in module Bio.Align.AlignInfo object:\n",
      "\n",
      "class SummaryInfo(builtins.object)\n",
      " |  SummaryInfo(alignment)\n",
      " |  \n",
      " |  Calculate summary info about the alignment.\n",
      " |  \n",
      " |  This class should be used to calculate information summarizing the\n",
      " |  results of an alignment. This may either be straight consensus info\n",
      " |  or more complicated things.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, alignment)\n",
      " |      Initialize with the alignment to calculate information on.\n",
      " |      \n",
      " |      ic_vector attribute. A list of ic content for each column number.\n",
      " |  \n",
      " |  dumb_consensus(self, threshold=0.7, ambiguous='X', require_multiple=False)\n",
      " |      Output a fast consensus sequence of the alignment.\n",
      " |      \n",
      " |      This doesn't do anything fancy at all. It will just go through the\n",
      " |      sequence residue by residue and count up the number of each type\n",
      " |      of residue (ie. A or G or T or C for DNA) in all sequences in the\n",
      " |      alignment. If the percentage of the most common residue type is\n",
      " |      greater then the passed threshold, then we will add that residue type,\n",
      " |      otherwise an ambiguous character will be added.\n",
      " |      \n",
      " |      This could be made a lot fancier (ie. to take a substitution matrix\n",
      " |      into account), but it just meant for a quick and dirty consensus.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - threshold - The threshold value that is required to add a particular\n",
      " |         atom.\n",
      " |       - ambiguous - The ambiguous character to be added when the threshold is\n",
      " |         not reached.\n",
      " |       - require_multiple - If set as True, this will require that more than\n",
      " |         1 sequence be part of an alignment to put it in the consensus (ie.\n",
      " |         not just 1 sequence and gaps).\n",
      " |  \n",
      " |  gap_consensus(self, threshold=0.7, ambiguous='X', require_multiple=False)\n",
      " |      Output a fast consensus sequence of the alignment, allowing gaps.\n",
      " |      \n",
      " |      Same as dumb_consensus(), but allows gap on the output.\n",
      " |      \n",
      " |      Things to do:\n",
      " |       - Let the user define that with only one gap, the result\n",
      " |         character in consensus is gap.\n",
      " |       - Let the user select gap character, now\n",
      " |         it takes the same as input.\n",
      " |  \n",
      " |  get_column(self, col)\n",
      " |      Return column of alignment.\n",
      " |  \n",
      " |  information_content(self, start=0, end=None, e_freq_table=None, log_base=2, chars_to_ignore=None, pseudo_count=0)\n",
      " |      Calculate the information content for each residue along an alignment.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - start, end - The starting an ending points to calculate the\n",
      " |         information content. These points should be relative to the first\n",
      " |         sequence in the alignment, starting at zero (ie. even if the 'real'\n",
      " |         first position in the seq is 203 in the initial sequence, for\n",
      " |         the info content, we need to use zero). This defaults to the entire\n",
      " |         length of the first sequence.\n",
      " |       - e_freq_table - A dictionary specifying the expected frequencies\n",
      " |         for each letter (e.g. {'G' : 0.4, 'C' : 0.4, 'T' : 0.1, 'A' : 0.1}).\n",
      " |         Gap characters should not be included, since these should not have\n",
      " |         expected frequencies.\n",
      " |       - log_base - The base of the logarithm to use in calculating the\n",
      " |         information content. This defaults to 2 so the info is in bits.\n",
      " |       - chars_to_ignore - A listing of characters which should be ignored\n",
      " |         in calculating the info content. Defaults to none.\n",
      " |      \n",
      " |      Returns:\n",
      " |       - A number representing the info content for the specified region.\n",
      " |      \n",
      " |      Please see the Biopython manual for more information on how information\n",
      " |      content is calculated.\n",
      " |  \n",
      " |  pos_specific_score_matrix(self, axis_seq=None, chars_to_ignore=None)\n",
      " |      Create a position specific score matrix object for the alignment.\n",
      " |      \n",
      " |      This creates a position specific score matrix (pssm) which is an\n",
      " |      alternative method to look at a consensus sequence.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - chars_to_ignore - A list of all characters not to include in\n",
      " |         the pssm.\n",
      " |       - axis_seq - An optional argument specifying the sequence to\n",
      " |         put on the axis of the PSSM. This should be a Seq object. If nothing\n",
      " |         is specified, the consensus sequence, calculated with default\n",
      " |         parameters, will be used.\n",
      " |      \n",
      " |      Returns:\n",
      " |       - A PSSM (position specific score matrix) object.\n",
      " |  \n",
      " |  replacement_dictionary(self, skip_chars=None, letters=None)\n",
      " |      Generate a replacement dictionary to plug into a substitution matrix.\n",
      " |      \n",
      " |      This should look at an alignment, and be able to generate the number\n",
      " |      of substitutions of different residues for each other in the\n",
      " |      aligned object.\n",
      " |      \n",
      " |      Will then return a dictionary with this information::\n",
      " |      \n",
      " |          {('A', 'C') : 10, ('C', 'A') : 12, ('G', 'C') : 15 ....}\n",
      " |      \n",
      " |      This also treats weighted sequences. The following example shows how\n",
      " |      we calculate the replacement dictionary. Given the following\n",
      " |      multiple sequence alignment::\n",
      " |      \n",
      " |          GTATC  0.5\n",
      " |          AT--C  0.8\n",
      " |          CTGTC  1.0\n",
      " |      \n",
      " |      For the first column we have::\n",
      " |      \n",
      " |          ('A', 'G') : 0.5 * 0.8 = 0.4\n",
      " |          ('C', 'G') : 0.5 * 1.0 = 0.5\n",
      " |          ('A', 'C') : 0.8 * 1.0 = 0.8\n",
      " |      \n",
      " |      We then continue this for all of the columns in the alignment, summing\n",
      " |      the information for each substitution in each column, until we end\n",
      " |      up with the replacement dictionary.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - skip_chars - Not used; setting it to anything other than None\n",
      " |         will raise a ValueError\n",
      " |       - letters - An iterable (e.g. a string or list of characters to include.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "align_info = AlignInfo.SummaryInfo(aligned)\n",
    "help(align_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The simple consensus: ACGTACGX\n",
      "Alignment score for particular bases at a given position.\n",
      "Dumb consensus on the y axis up-to-down, bases on the x-axis.\n",
      "    A   C   G   T\n",
      "A  6.0 2.0 0.0 0.0\n",
      "C  0.0 5.0 1.0 1.0\n",
      "G  0.0 0.0 8.0 0.0\n",
      "T  0.0 0.0 0.0 7.0\n",
      "A  7.0 0.0 1.0 0.0\n",
      "C  0.0 6.0 0.0 0.0\n",
      "G  0.0 1.0 5.0 0.0\n",
      "X  0.0 2.0 1.0 3.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"The simple consensus:\", align_info.dumb_consensus())\n",
    "print(\"Alignment score for particular bases at a given position.\")\n",
    "print(\"Dumb consensus on the y axis up-to-down, bases on the x-axis.\")\n",
    "print(align_info.pos_specific_score_matrix())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "## Downloading and reading FASTA files\n",
    "\n",
    "FASTA is a simple, plain-text-based sequence format for representing sequences, which supports multiple records contained in the same file. An example:\n",
    "\n",
    ">\\> MCHU - Calmodulin - Human, rabbit, bovine, rat, and chicken\n",
    "MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID\n",
    "FPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREA\n",
    "DIDGDGQVNYEEFVQMMTAK*\n",
    "\n",
    "We can see that in addition to the raw aminoacid sequence there is some metadata denoting where the sequence came from and what it represents.\n",
    "\n",
    "For our example, we will be using [CFTR gene](https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000001626;r=7:117287120-117715971), encoding [a protein](https://www.uniprot.org/uniprot/P13569).\n",
    "\n",
    "First, let's download ther relevant FASTA file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded file to:  data/ENSG00000001626.fasta\n",
      "\n",
      "Download metadata and statistics:\n",
      "Vary: Content-Type\n",
      "Vary: Origin\n",
      "Content-Type: text/x-fasta; charset=UTF-8\n",
      "Date: Tue, 23 Nov 2021 18:45:51 GMT\n",
      "X-RateLimit-Limit: 55000\n",
      "X-RateLimit-Reset: 849\n",
      "X-Runtime: 1.262950\n",
      "Connection: close\n",
      "X-RateLimit-Period: 3600\n",
      "X-RateLimit-Remaining: 54998\n",
      "Content-Length: 436062\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from urllib.request import urlretrieve \n",
    "\n",
    "CFTR_FASTA_path = \"data/ENSG00000001626.fasta\"\n",
    "result_location, http_response = urlretrieve(\"https://rest.ensembl.org/sequence/id/ENSG00000001626.fasta\", CFTR_FASTA_path)\n",
    "print(\"Downloaded file to: \", result_location)\n",
    "\n",
    "print(\"\\nDownload metadata and statistics:\")\n",
    "print(http_response)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "The above message may look complicated - it contains a lot of information related to the transfer and the type of the file. \n",
    "\n",
    "To open the file and read relevant information from it, we use `SeqIO` module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total records extracted: 1\n",
      "The first record:        ID: ENSG00000001626.16\n",
      "Name: ENSG00000001626.16\n",
      "Description: ENSG00000001626.16 chromosome:GRCh38:7:117287120:117715971:1\n",
      "Number of features: 0\n",
      "Seq('AGGCGGATCACAAGTTCATGAGATCGAGACCATCTTGGCCAACATGGTGAGACC...ACA')\n",
      "Type of records:         <class 'Bio.SeqRecord.SeqRecord'>\n",
      "Help on SeqRecord in module Bio.SeqRecord object:\n",
      "\n",
      "class SeqRecord(builtins.object)\n",
      " |  SeqRecord(seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |  \n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |   - id          - Identifier such as a locus tag (string)\n",
      " |   - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |   - name        - Sequence name, e.g. gene name (string)\n",
      " |   - description - Additional text (string)\n",
      " |   - dbxrefs     - List of database cross references (list of strings)\n",
      " |   - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |   - annotations - Further information about the whole sequence (dictionary).\n",
      " |     Most entries are strings, or lists of strings.\n",
      " |   - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |     dictionary). This holds Python sequences (lists, strings\n",
      " |     or tuples) whose length matches that of the sequence.\n",
      " |     A typical use would be to hold a list of integers\n",
      " |     representing sequencing quality scores, or a string\n",
      " |     representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Define the equal-to operand (not implemented).\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the Python format() function and f-strings.\n",
      " |      The format_spec should be a lower case string supported by\n",
      " |      Bio.SeqIO as a text output file format. Requesting a binary file\n",
      " |      format raises a ValueError. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      ...\n",
      " |      >>> format(record, \"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(f\"Here is {record.id} in FASTA format:\\n{record:fasta}\")\n",
      " |      Here is YP_025292.1 in FASTA format:\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      See also the SeqRecord's format() method.\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Define the greater-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with some annotation preserved as follows:\n",
      " |      \n",
      " |      * The name, id and description are kept as-is.\n",
      " |      * Any per-letter-annotations are sliced to match the requested\n",
      " |        sub-sequence.\n",
      " |      * Unless a stride is used, all those features which fall fully\n",
      " |        within the subsequence are included (with their locations\n",
      " |        adjusted accordingly). If you want to preserve any truncated\n",
      " |        features (e.g. GenBank/EMBL source features), you must\n",
      " |        explicitly add them to the new SeqRecord yourself.\n",
      " |      * With the exception of any molecule type, the annotations\n",
      " |        dictionary and the dbxrefs list are not used for the new\n",
      " |        SeqRecord, as in general they may not apply to the\n",
      " |        subsequence. If you want to preserve them, you must explicitly\n",
      " |        copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\"),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"]\n",
      " |      '  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  '\n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD')\n",
      " |      >>> sub.letter_annotations[\"secondary_structure\"]\n",
      " |      'HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH'\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD')\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR')\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Define the greater-than operand (not implemented).\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq         - Sequence, required (Seq or MutableSeq)\n",
      " |       - id          - Sequence identifier, recommended (string)\n",
      " |       - name        - Sequence name, optional (string)\n",
      " |       - description - Sequence description, optional (string)\n",
      " |       - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |       - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |       - annotations - Dictionary of annotations for the whole sequence\n",
      " |       - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |         should be strings, list or tuples of the same length as the full\n",
      " |         sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __le___(self, other)\n",
      " |      Define the less-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Define the less-than operand (not implemented).\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Define the not-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                     \"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                     \"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                     \"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\"),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The Python print function automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only,\n",
      " |      and binary formats are not supported).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Return new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the reversed sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement, here we explicitly give a new\n",
      " |      identifier (the old identifier with a suffix):\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id + \"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG')\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC')\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\"), id=\"Test\",\n",
      " |      ...                         annotations={\"molecule_type\": \"protein\"})\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      If you have RNA without any U bases, it must be annotated as RNA\n",
      " |      otherwise it will be treated as DNA by default with A mapped to T:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rna1 = SeqRecord(Seq(\"ACG\"), id=\"Test\")\n",
      " |      >>> rna2 = SeqRecord(Seq(\"ACG\"), id=\"Test\", annotations={\"molecule_type\": \"RNA\"})\n",
      " |      >>> print(rna1.reverse_complement(id=\"RC\", description=\"unk\").format(\"fasta\"))\n",
      " |      >RC unk\n",
      " |      CGT\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rna2.reverse_complement(id=\"RC\", description=\"RNA\").format(\"fasta\"))\n",
      " |      >RC RNA\n",
      " |      CGU\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\"), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None, id=False, name=False, description=False, features=False, annotations=False, letter_annotations=False, dbxrefs=False)\n",
      " |      Return new SeqRecord with translated sequence.\n",
      " |      \n",
      " |      This calls the record's .seq.translate() method (which describes\n",
      " |      the translation related arguments, like table for the genetic code),\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the translated sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or False (default) to omit them.\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent\n",
      " |      (annotations only), or False (default) to omit them.\n",
      " |      \n",
      " |      e.g. Loading a FASTA gene and translating it,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> gene_record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> print(gene_record.format(\"fasta\"))\n",
      " |      >gi|3176602|gb|U78617.1|LOU78617 Lathyrus odoratus phytochrome A (PHYA) gene, partial cds\n",
      " |      CAGGCTGCGCGGTTTCTATTTATGAAGAACAAGGTCCGTATGATAGTTGATTGTCATGCA\n",
      " |      AAACATGTGAAGGTTCTTCAAGACGAAAAACTCCCATTTGATTTGACTCTGTGCGGTTCG\n",
      " |      ACCTTAAGAGCTCCACATAGTTGCCATTTGCAGTACATGGCTAACATGGATTCAATTGCT\n",
      " |      TCATTGGTTATGGCAGTGGTCGTCAATGACAGCGATGAAGATGGAGATAGCCGTGACGCA\n",
      " |      GTTCTACCACAAAAGAAAAAGAGACTTTGGGGTTTGGTAGTTTGTCATAACACTACTCCG\n",
      " |      AGGTTTGTT\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      And now translating the record, specifying the new ID and description:\n",
      " |      \n",
      " |      >>> protein_record = gene_record.translate(table=11,\n",
      " |      ...                                        id=\"phya\",\n",
      " |      ...                                        description=\"translation\")\n",
      " |      >>> print(protein_record.format(\"fasta\"))\n",
      " |      >phya translation\n",
      " |      QAARFLFMKNKVRMIVDCHAKHVKVLQDEKLPFDLTLCGSTLRAPHSCHLQYMANMDSIA\n",
      " |      SLVMAVVVNDSDEDGDSRDAVLPQKKKRLWGLVVCHNTTPRFV\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\"), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |      \n",
      " |      Note that if replacing the record's sequence with a sequence of a\n",
      " |      different length you must first clear the letter_annotations dict.\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Reading FASTA files. Note that each file can contain multiple records.\n",
    "from Bio import SeqIO\n",
    "records = [r for r in SeqIO.parse(CFTR_FASTA_path, \"fasta\")]\n",
    "print(\"Total records extracted:\", len(records))\n",
    "print(\"The first record:       \", records[0])\n",
    "print(\"Type of records:        \", type(records[0]))\n",
    "help(records[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "We can see that `SeqRecord`, a class used to represent items read from FASTA files is relatively big and contains placeholders for a lot of possible information, including references, description, annotations etc.\n",
    "\n",
    "To get the actual `Seq` object we access `seq` property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 428852\n",
      "GC content: 36.895245912342716\n"
     ]
    }
   ],
   "source": [
    "record = records[0]\n",
    "print(\"Sequence length:\", len(record.seq))\n",
    "print(\"GC content:\", GC(record.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": false
   },
   "source": [
    "## Discussion\n",
    "\n",
    "In this notebook we have discussed the basic ways sequences, the bread-and-butter of bioinformatics, can be constructed and operated on, including allowing for transcription and translation. We have seen  how **Biopython** supports sequence aligning - both pairwise and MSA. We have looked at how to obtain and open FASTA files to retrieve the records with associated metadata. \n",
    "\n",
    "This notebook only touches on the tools and integrations provided with **Biopython** library, the plethora of formats and tools that can be used to store and process sequences and alignment data.\n",
    "\n",
    "This notebook relied on very simple toy sequences. In practice, this is unrealistic. Try going through this notebook using longer, more representative sequences and experimenting with the code.\n",
    "\n",
    "The exercise section that accompanies this notebook will allow you to use the code here in more practical scenarios."
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
