{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MEDC0106: Bioinformatics in Applied Biomedical Science\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"../../resources/static/Banner.png\" alt=\"MEDC0106 Banner\" width=\"90%\"/>\n",
    "  <br>\n",
    "</p>\n",
    "\n",
    "---------------------------------------------------------------\n",
    "\n",
    "# 04 - Session 1 Exercises\n",
    "\n",
    "*Written by:* Oliver Scott\n",
    "\n",
    "**This notebook contains exercises based around the material provided in session one**\n",
    "\n",
    "Try to complete the exercises provided before the next session, feel free to refer back to the sessions content to help you complete the tasks.\n",
    "\n",
    "You should work through the tasks consecutively.\n",
    "\n",
    "Save your results so we can go through the answers in the next session.\n",
    "\n",
    "----\n",
    "\n",
    "## Contents\n",
    "\n",
    "1. [Task 1](#Task-1) - Starting Basic\n",
    "2. [Task 2](#Task-2) - Nucleotide Counting\n",
    "3. [Task 3](#Task-3) - Sequence Translation\n",
    "4. [Task 4](#Task-4) - Sequence Mutation\n",
    "5. [Task 5](#Task-5) - Rock, Paper, Scissors\n",
    "\n",
    "----\n",
    "\n",
    "## Task 1\n",
    "\n",
    "#### Starting Basic\n",
    "\n",
    "A scientist in your lab has conducted a survey asking members of the public if they think the research they are doing will benefit them. They have asked you to calculate some basic statistics for them:\n",
    "\n",
    "- The total number of participants\n",
    "- The mean age of the participants\n",
    "- The proportion of male ('M') to female ('F') participants\n",
    "- The percentage of 'yes' ('Y') answers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The scientists survey results:\n",
    "\n",
    "age = [45,35,37,49,31,45,43,30,48,36,38,44,39,36,40,42,38,30,31,39]\n",
    "gender = ['F','F','F','M','F','F','F','M','F','M','F','F','F','M','F','F','F','M','F','F']\n",
    "answer = ['Y','N','N','Y','Y','Y','N','N','N','N','Y','N','N','N','Y','N','Y','N','N','Y']\n",
    "\n",
    "# Calculate the required statistics for the scientist and print them below.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2\n",
    "\n",
    "#### Nucleotide Counting\n",
    "\n",
    "In this task we have given you a probe nucleotide sequence and some incomplete code. We would like you to complete the function `count_atcg` so that it returns the number of each nucleotide (A, T, C, G) in the sequence as a dictionary. The first step has been done for you.\n",
    "\n",
    "<details>\n",
    "<summary>Click here for a hint!</summary>\n",
    "<em>Strings can be iterated just like lists with a simple for loop!</em>\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete this function to complete the task!\n",
    "\n",
    "def count_atcg(sequence):\n",
    "    counts = {'A': 0, 'T': 0, 'C': 0, 'G': 0}\n",
    "    # \n",
    "    #\n",
    "    # Complete the function to return the counts\n",
    "    #\n",
    "    #\n",
    "    return counts\n",
    "\n",
    "\n",
    "# Ignore the backslashes '\\', they wont appear when iterating the string\n",
    "probe_sequence = \"\"\"\\\n",
    "ATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAG\\\n",
    "CCGCAGCCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTG\\\n",
    "CGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGG\\\n",
    "CAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCC\\\n",
    "TGCAGAAGCGTGGCATTGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAA\\\n",
    "CTACTGCAAC\"\"\"\n",
    "\n",
    "atcg_content = count_atcg(probe_sequence)\n",
    "\n",
    "print('Counts for probe sequence:')\n",
    "print('A:', atcg_content['A'])\n",
    "print('T:', atcg_content['T'])\n",
    "print('C:', atcg_content['C'])\n",
    "print('G:', atcg_content['G'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3\n",
    "\n",
    "#### Sequence Translation (harder)\n",
    "\n",
    "In this task we would like you to translate the above `probe_sequence` (Task ..) into an amino acid sequence using a codon table. The codon table below ([NCBI standard](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?chapter=tgencodes#SG1)) has been provided to you in the form of a Python dictionary. Complete the `translate` function below to return the fully translated sequence.\n",
    "\n",
    "The [range function](https://www.w3schools.com/python/ref_func_range.asp) `range(first, last, step)` might come in handy!  \n",
    "\n",
    "After you have translated the sequence use web tools to identify the protein!\n",
    "\n",
    "##### NCBI Codon Table:\n",
    "```\n",
    "TTT F Phe      TCT S Ser      TAT Y Tyr      TGT C Cys  \n",
    "TTC F Phe      TCC S Ser      TAC Y Tyr      TGC C Cys  \n",
    "TTA L Leu      TCA S Ser      TAA * Ter      TGA * Ter  \n",
    "TTG L Leu i    TCG S Ser      TAG * Ter      TGG W Trp  \n",
    "\n",
    "CTT L Leu      CCT P Pro      CAT H His      CGT R Arg  \n",
    "CTC L Leu      CCC P Pro      CAC H His      CGC R Arg  \n",
    "CTA L Leu      CCA P Pro      CAA Q Gln      CGA R Arg  \n",
    "CTG L Leu i    CCG P Pro      CAG Q Gln      CGG R Arg  \n",
    "\n",
    "ATT I Ile      ACT T Thr      AAT N Asn      AGT S Ser  \n",
    "ATC I Ile      ACC T Thr      AAC N Asn      AGC S Ser  \n",
    "ATA I Ile      ACA T Thr      AAA K Lys      AGA R Arg  \n",
    "ATG M Met i    ACG T Thr      AAG K Lys      AGG R Arg  \n",
    "\n",
    "GTT V Val      GCT A Ala      GAT D Asp      GGT G Gly  \n",
    "GTC V Val      GCC A Ala      GAC D Asp      GGC G Gly  \n",
    "GTA V Val      GCA A Ala      GAA E Glu      GGA G Gly  \n",
    "GTG V Val      GCG A Ala      GAG E Glu      GGG G Gly\n",
    "```\n",
    "***Note:***  _* or 'Ter' indicates a termination codon_\n",
    "\n",
    "- *Try not to use the hints unless you really need to*\n",
    "- *Remember that you can use assignment operators with strings `+=`*\n",
    "\n",
    "<details>\n",
    "<summary>Click here for a hint</summary>\n",
    "<em>The range(first, last, step) function can be used with a for loop to iterate over a range of values with an optional step</em>\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary>Click here for another hint</summary>\n",
    "<em>Strings can be sliced like lists (sequence[0:3])</em>\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary>Click here for another hint</summary>\n",
    "<em>\"for ix in range(0, len(sequence), 3)\"</em>\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary>Click here for a final hint</summary>\n",
    "<em>\"sequence[ix : ix + 3]\"</em>\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_codon_table = {\n",
    "    'TTT': 'F', 'TCT': 'S', 'TAT': 'Y', 'TGT': 'C',\n",
    "    'TTC': 'F', 'TCC': 'S', 'TAC': 'Y', 'TGC': 'C',\n",
    "    'TTA': 'L', 'TCA': 'S', 'TAA': '*', 'TGA': '*',\n",
    "    'TTG': 'L', 'TCG': 'S', 'TAG': '*', 'TGG': 'W',\n",
    "    'CTT': 'L', 'CCT': 'P', 'CAT': 'H', 'CGT': 'R',\n",
    "    'CTC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',\n",
    "    'CTA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',\n",
    "    'CTG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',\n",
    "    'ATT': 'I', 'ACT': 'T', 'AAT': 'N', 'AGT': 'S',\n",
    "    'ATC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',\n",
    "    'ATA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',\n",
    "    'ATG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',\n",
    "    'GTT': 'V', 'GCT': 'A', 'GAT': 'D', 'GGT': 'G',\n",
    "    'GTC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',\n",
    "    'GTA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',\n",
    "    'GTG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "\n",
    "# Complete this function to complete the task!\n",
    "def translate(sequence):\n",
    "    translated = \"\"\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    # Complete the function to return the translation\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    return translated\n",
    "\n",
    "\n",
    "aa_sequence = translate(probe_sequence)\n",
    "print('Translated:', aa_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 4\n",
    "\n",
    "#### Sequence Mutation\n",
    "\n",
    "Why would a mutation from 'G' to 'A' at position 12 in the probe sequence likely cause disease? You should be able to create some code to find out!\n",
    "\n",
    "- Complete the mutation function to mutate the nucleotide at position 12 (returning a new sequence)\n",
    "- Use the translate function you defined in the exercise to translate the mutated sequence\n",
    "- Work out why this mutation may cause disease\n",
    "\n",
    "To clarify position 12 means the 12th position in the sequence not the 12th index\n",
    "\n",
    "***Note:*** *Remember that indexing in Python starts at 0!*\n",
    "\n",
    "\n",
    "<details>\n",
    "<summary>Click here for a hint</summary>\n",
    "<em>Strings are immutable/unchangeable so you will need to create a new string</em>\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary>Click here for another hint</summary>\n",
    "<em>Strings are sliceable! i.e. sequence[:i - 1]</em>\n",
    "</details>\n",
    "\n",
    "<details>\n",
    "<summary>Click here for a final hint</summary>\n",
    "<em>Strings can be concatanated with '+'. i.e.  new_string = sequence[:i-] + mutation</em>\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_sequence(sequence, position, mutation):\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    # Complete this function to mutate the sequence\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    return \n",
    "\n",
    "\n",
    "mutated_sequence = mutate_sequence(probe_sequence, 12, 'A')\n",
    "\n",
    "# Now use your translate function to see why this mutation may cause disease!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 5\n",
    "\n",
    "#### Rock, Paper, Scissors\n",
    "\n",
    "In this task we would like you to complete the rock, paper, scissors game defined in `mymodule.py`. You should complete the `determine_winner` function in mymodule.py, printing the winner of the game ('Player Wins!', 'Computer Wins!' or 'Draw!') to the screen using the `print` function:\n",
    "\n",
    "If you are not familiar with the rules of the game:\n",
    "\n",
    "- Rock beats Scissors\n",
    "- Scissors beats Paper\n",
    "- Paper beats Rock\n",
    "- Same choice results in a draw\n",
    "\n",
    "When the cell below is run it will call the `play_game` function defined in mymodule which will call the completed `determine_winner` function to print the results to the screen. When you run the code you will need to type in your choice in the box that appears!\n",
    "\n",
    "You do **Not** need to edit any code in the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "outputs": [],
   "source": [
    "import sys; sys.modules.pop('mymodule', None)  # Ignore this line (refreshes mymodule)\n",
    "\n",
    "import mymodule\n",
    "\n",
    "mymodule.play_game()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "Remember to save your answers! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
